domains/client/domain-executor/src/bundle_processor.rs

```diff
@@ -149,9 +150,36 @@ where
             let mut domain_parent = initial_parent;
 
             for primary_info in primary_imports {
-                domain_parent = self
+                if let Some(next_domain_parent) = self
                     .process_bundles_at((primary_info.hash, primary_info.number), domain_parent)
+                    .await?
+                {
+                    domain_parent = next_domain_parent;
+                }
+            }
+
+            // The domain branch driving from the best primary branch should also be the best domain branch even
```

Comment:
This edge case shouldn't persist if we pass the fork choice of consensus block in the notification instead of using the LongestChain on the domain block, let's add a TODO just in case it's possible someday (we had it before, when using the subspace block import notification, but lost it when switching to the substrate block import notification).

```suggestion
            // TODO: Remove once the fork choice of consensus block can be directly applied on the domain block.
            // 
            // The domain branch derived from the best primary branch should also be the best domain branch even
```

Reply:
We may still need this, because a consensus block may not drive a domain block, and as long as the incoming consensus block is the new best block, then the best fork of the domain chain may need to change even if no new domain block is produced/imported.

```
      b2 -> b3 -> b4
     /
a1 -> a2 -> a3
```

```
            domain_b2
          /
domain_a1 -> domain_a2 -> domain_a3
```

Take the above as an example, initially the best block of the consensus chain and domain chain was `a3` and `domain_a3` respectively. After `b4` is imported, fork B becomes the best fork of the consensus chain, and the best fork of the domain chain also changes to fork B even though there is no new domain block produced in fork B due to `b3` and `b4` don't contain any bundle.

Reply:
I meant to apply the fork choice of every consensus chain block to the corresponding domain block. In this case, a3 -> domain_a3, when the consensus chain switches from a3 to b4, b2 -> domain_b2, b3 -> domain_b2, b4 -> domain_b2, so in the end, the fork choice of b4 will be applied on domain_b2, the domain chain is switched from domain_a3 to domain_b2 naturally.

Hmm, on second thought, I think re-applying every consensus block's fork choice technically works the same as here (using the longest chain in the domain block building and re-import the last as best), but it feels more straightforward to understand IMO. Let's keep it as is then.